### PR TITLE
Feature/docker pre built openwrt

### PR DESCRIPTION
**Pull request by rmelotte**
This PR adds a Dockerfile and scripts to cross-compile prplMesh for OpenWrt target.

Although it should be possible to use it for the rax40 platform, it's currently not supported due to wrong dependencies in the package Makefiles.
We will have to fix the existing prplWRT feeds first.

### CODE REVIEW

**Comment by rmelotte** [comment]
DO NOT MERGE
This needs to be updated to use another prplmesh feed.

**Comment by arnout** [comment]
[Build for target as part of CI](https://trello.com/c/AoN6HvdK/30-build-for-target-as-part-of-ci)

File: tools/docker/builder/openwrt/Dockerfile
Comment (arnout) [inline tools/docker/builder/openwrt/Dockerfile]: I'd like some documentation comments above the ARGs.

File: tools/docker/builder/openwrt/Dockerfile
Comment (arnout) [inline tools/docker/builder/openwrt/Dockerfile]: Since sdk_config is just a oneliner, and we definitely always want that one line in the config, I'd do it with echo here instead of requiring a separate file. If it becomes more than that one line, we can still make it a file.

File: tools/docker/builder/openwrt/Dockerfile
Comment (arnout) [inline tools/docker/builder/openwrt/Dockerfile]: I guess this is for debugging. However, I don't think it's that useful, it really can't be anything else than those two lines, right? Anyway, if it is kept, there should also be an echo before to tell us what we're seeing.

File: tools/docker/builder/openwrt/Dockerfile
Comment (arnout) [inline tools/docker/builder/openwrt/Dockerfile]: Not sure if tee'ing to build.log makes a whole lot of sense, but OK.

File: tools/docker/builder/openwrt/Dockerfile
Comment (arnout) [inline tools/docker/builder/openwrt/Dockerfile]: Why `-exec echo {} \;` instead of the normal `-print`?

File: tools/docker/builder/openwrt/Dockerfile
Comment (arnout) [inline tools/docker/builder/openwrt/Dockerfile]: Please use a different variable name, reusing the same one is just confusing.

However, all this cd and rm stuff is redundant, you can just do

```
    && mkdir -p "${HOME}/openwrt_sdk" \
    && tar -C "${HOME}/openwrt_sdk" -xf "${TAR}" --strip-components=1
```

File: tools/docker/builder/openwrt/Dockerfile
Comment (arnout) [inline tools/docker/builder/openwrt/Dockerfile]: Is it useful to make this an argument?

If yes, please document it (including how it could be used).

File: tools/docker/builder/openwrt/Dockerfile
Comment (arnout) [inline tools/docker/builder/openwrt/Dockerfile]: That's what I'm talking about: document the ARGs!

File: tools/docker/builder/openwrt/Dockerfile
Comment (arnout) [inline tools/docker/builder/openwrt/Dockerfile]: I would put all this in the base image already. Although I wonder why less, vim and libgetopt-complete are useful..

File: tools/docker/builder/openwrt/Dockerfile
Comment (arnout) [inline tools/docker/builder/openwrt/Dockerfile]: If you do this at the beginning, you don't need the additional update below.

File: tools/docker/builder/openwrt/Dockerfile
Comment (arnout) [inline tools/docker/builder/openwrt/Dockerfile]: Since this is a hopefully temporary hack, I'd do it in a separate RUN. And add a comment to explain this hack.

TBH I don't understand why the feeds install has to be done after the patching. AFAIU it should be possible to locally modify a Makefile without re-running feeds.

File: tools/docker/builder/openwrt/build.sh
Comment (arnout) [inline tools/docker/builder/openwrt/build.sh:69]: I don't like this, but it's OK for the time being.

What we really want is that the current source gets built inside the container, not some random commit that you must have pushed already and that you have to provide explicitly on the command line.

A first step to achieve that is pretty simple: mount the .git directory inside the container as /home/openwrt/.../prplmesh/git-src. Then we just require the current source to be checked in as HEAD.

**Review by arnout — CHANGES_REQUESTED** [review]
Lots of small things.

File: tools/docker/builder/openwrt/Dockerfile
Comment (rmelotte) [reply tools/docker/builder/openwrt/Dockerfile]: Done.

File: tools/docker/builder/openwrt/Dockerfile
Comment (rmelotte) [reply tools/docker/builder/openwrt/Dockerfile]: It's not useful anymore indeed. I removed it.

File: tools/docker/builder/openwrt/Dockerfile
Comment (rmelotte) [reply tools/docker/builder/openwrt/Dockerfile]: No particular reason, I switched to `-print`.

File: tools/docker/builder/openwrt/Dockerfile
Comment (rmelotte) [reply tools/docker/builder/openwrt/Dockerfile]: Indeed, that's so much simpler. I also skipped the "rm" part entirely, since there is no real reason to delete it (from the `openwrt` dir; it won't be in `openwrt_sdk`).

File: tools/docker/builder/openwrt/Dockerfile
Comment (rmelotte) [reply tools/docker/builder/openwrt/Dockerfile]: If we ever need to add a new build-time dependency (one that could reasonably taken from the host) or add other debugging tools, we could do so without recompiling the base image (the `openwrt-prebuilt` one takes significantly more time to build than the other ones).

`less`, `vim` et al. have been added to be able to debug failing prplMesh builds (when running the container of course, not when buiding) interactively. 
I don't remember why `libgetopt-complete` though. I'll have to check if it's still used and either remove it or add a comment.

On the other hand it's not like we can keep the same base image forever, we will have to update it from time to time anyway.

What do you think? Is it worth keeping these packages in a separate step?

Comment (arnout) [reply tools/docker/builder/openwrt/Dockerfile]: > If we ever need to add a new build-time dependency (one that could reasonably taken from the host) or add other debugging tools, we could do so without recompiling the base image (the `openwrt-prebuilt` one takes significantly more time to build than the other ones).

We can still use this approach when we need that. But I think even then we'll at some point consolidate it into the base image.

> `less`, `vim` et al. have been added to be able to debug failing prplMesh builds (when running the container of course, not when buiding) interactively.

OK, good reason to keep them. Add a comment about it though.

File: tools/docker/builder/openwrt/Dockerfile
Comment (rmelotte) [reply tools/docker/builder/openwrt/Dockerfile]: Done.

File: tools/docker/builder/openwrt/Dockerfile
Comment (rmelotte) [reply tools/docker/builder/openwrt/Dockerfile]: Done.

File: tools/docker/builder/openwrt/Dockerfile
Comment (rmelotte) [reply tools/docker/builder/openwrt/Dockerfile]: Actually this is not needed anymore, I removed it.

File: tools/docker/builder/openwrt/Dockerfile
Comment (rmelotte) [reply tools/docker/builder/openwrt/Dockerfile]: Indeed it was most useful when debugging the first (successful) builds, removing.

File: tools/docker/builder/openwrt/Dockerfile
Comment (rmelotte) [reply tools/docker/builder/openwrt/Dockerfile]: Alright, I moved them to the base image.
NB: `libgetopt-complete-perl` was not needed anymore.

File: tools/docker/builder/openwrt/Dockerfile
Comment (arnout) [inline tools/docker/builder/openwrt/Dockerfile]: Could you add a documentation comment for these three as well?

File: tools/docker/builder/openwrt/configs/sdk_config
Comment (arnout) [inline tools/docker/builder/openwrt/configs/sdk_config]: This file should be removed now.

File: tools/docker/builder/openwrt/scripts/build.sh
Comment (arnout) [inline tools/docker/builder/openwrt/scripts/build.sh]: Is there a chance that the find will match several ipk files?

**Review by arnout — CHANGES_REQUESTED** [review]

File: tools/docker/builder/openwrt/configs/sdk_config
Comment (rmelotte) [reply tools/docker/builder/openwrt/configs/sdk_config]: Done.

File: tools/docker/builder/openwrt/Dockerfile
Comment (rmelotte) [reply tools/docker/builder/openwrt/Dockerfile]: Not really useful no, it has been removed.

File: tools/docker/builder/openwrt/Dockerfile
Comment (rmelotte) [reply tools/docker/builder/openwrt/Dockerfile]: Done.

File: tools/docker/builder/openwrt/scripts/build.sh
Comment (rmelotte) [reply tools/docker/builder/openwrt/scripts/build.sh]: In theory yes, if another package builds an ipk matching "prplmesh*".
I don't see any easy way to get an exact match at this point though. 
We could check that the version number is following the package name maybe.
For reference, here is the full package name for the rax40:  `prplmesh_1.0-1_mips_24kc_nomips16.ipk`.

File: tools/docker/builder/openwrt/build.sh
Comment (rmelotte) [reply tools/docker/builder/openwrt/build.sh:69]: I went with [USE_SOURCE_DIR](https://openwrt.org/docs/guide-developer/packages#use_source_dir) since it doesn't require to commit changes in order to build. As it requires a separate "prepare" step, we will have to document how to rebuild manually  somewhere.

**Comment by rmelotte** [comment]
Rebased on master.
Also made the git describe output shorter, so that the "prplmesh version" part of the resulting .ipk is only a hash (and potentially "-dirty").

File: tools/docker/builder/openwrt/build.sh
Comment (arnout) [inline tools/docker/builder/openwrt/build.sh:10]: AFAICS there are now defaults for everything, so it is not actually needed to pass a config file.

However, I think it's nicer to use command line arguments instead of a config file, using getopt, like we do in most other scripts.

This is something we can easily update later so (nitpick).

**Review by arnout — APPROVED** [review]
One nitpick. Please rebase on master so we can merge.

File: tools/docker/builder/openwrt/build.sh
Comment (rmelotte) [reply tools/docker/builder/openwrt/build.sh:10]: Indeed there are now defaults for everything. Now that you mention it, I should remove the default for TARGET and TARGET_PLATFORM_TYPE, as this build script will be used to build for other targets as well.
The idea of the config file was to make it easy for anyone to build for a specific target, without the need to know what TARGET and TARGET_PLATFORM_TYPE should be.

What I could do instead is to add a mandatory argument for the target and list the targets in the same script (currently only turris-omnia).
I'll rebase on master and keep it like that for now, please ping me if you want that done before merging.

**Pull request merged**
